File: code/ahasmove/106.从中序与后序遍历序列构造二叉树.cpp
/*
 * @lc app=leetcode.cn id=106 lang=cpp
 *
 * [106] 从中序与后序遍历序列构造二叉树
 *
 * https://leetcode-cn.com/problems/construct-binary-tree-from-inorder-and-postorder-traversal/description/
 *
 * algorithms
 * Medium (70.78%)
 * Likes:    405
 * Dislikes: 0
 * Total Accepted:    77.6K
 * Total Submissions: 109.7K
 * Testcase Example:  '[9,3,15,20,7]\n[9,15,7,20,3]'
 *
 * 根据一棵树的中序遍历与后序遍历构造二叉树。
 * 
 * 注意:
 * 你可以假设树中没有重复的元素。
 * 
 * 例如，给出
 * 
 * 中序遍历 inorder = [9,3,15,20,7]
 * 后序遍历 postorder = [9,15,7,20,3]
 * 
 * 返回如下的二叉树：
 * 
 * ⁠   3
 * ⁠  / \
 * ⁠ 9  20
 * ⁠   /  \
 * ⁠  15   7
 * 
 * 
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

// struct TreeNode
// {
//     int val;
//     TreeNode *left;
//     TreeNode *right;
//     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
// };

#include <vector>

using namespace std;

class Solution
{
public:
    TreeNode *div(vector<int> &inorder, int l1, int r1, vector<int> &postorder, int l2, int r2)
    {
        if(r2 < l2)
            return nullptr;

        int midVal = postorder[r2];
        int mid = l1;
        while (inorder[mid] != midVal)
            mid++;

        TreeNode *node = new TreeNode(midVal);

        node->left = div(inorder, l1, mid - 1, postorder, l2, l2 + mid - l1 - 1);
        node->right = div(inorder, mid + 1, r1, postorder, l2 + mid - l1, r2-1);

        return node;
    }

    TreeNode *buildTree(vector<int> &inorder, vector<int> &postorder)
    {
        
        return div(inorder, 0, inorder.size() - 1, postorder, 0, postorder.size() - 1);
    }
};
// @lc code=end

File: code/JavaLab/src/main/java/algorithm/leetcode/tree/106.从中序与后序遍历序列构造二叉树.java
package algorithm.leetcode.tree;

import algorithm.config.TreeNode;

/**
 * https://leetcode-cn.com/problems/construct-binary-tree-from-inorder-and-postorder-traversal/
 * <p>
 * 根据一棵树的中序遍历与后序遍历构造二叉树。
 */
class LC106 {

    public TreeNode buildTree(int[] inorder, int[] postorder) {
        return helper(inorder, 0, inorder.length - 1, postorder, 0, postorder.length - 1);
    }

    public TreeNode helper(int[] inorder, int l1, int r1, int[] postorder, int l2, int r2) {

        if (l1 > r1) {
            return null;
        }

        // 找到根节点的值
        int rootVal = postorder[r2];
        TreeNode root = new TreeNode(rootVal);

        // 从中序遍历中找到根节点下标
        int index = l1;
        while (inorder[index] != rootVal)
            index++;
        int diff = index - l1;

        root.left = helper(inorder, l1, index - 1, postorder, l2, l2 + diff - 1);
        root.right = helper(inorder, index + 1, r1, postorder, l2 + diff, r2 - 1);
        return root;
    }

}